grr/server/grr_response_server/gui/ui/components/flow_details/helpers/osquery_test_utils.ts

```diff
@@ -0,0 +1,208 @@
+    }
+
+    this.table.query = query;
+
+    return this;
+  }
+
+  withQueryIfDefined(query: string): OsqueryTableBuilder {
+    if (this.table !== undefined) {
+      this.table.query = query;
+    }
+
+    return this;
+  }
+
+  build(): FullMutableOsqueryTable | undefined {
+    return this.table;
+  }
+
+  buildWithQueryIfDefined(query: string): FullMutableOsqueryTable | undefined {
+    this.withQueryIfDefined(query);
+    return this.build();
+  }
+}
+
+/** Helper class to build a FlowListEntry objects in a declarative manner */
+export class OsqueryFlowListEntryBuilder {
+  private flowListEntry = newFlowListEntry({args: {query: ''}})  as DeepMutable<FlowListEntry>;
+
+  private query = '';
+
+  private stderr = '';
+  private resultsTableBuilder = new OsqueryTableBuilder();
+
+  private progressRowsCount = 0;
+  private progressTableBuilder = new OsqueryTableBuilder();
+
+  withFlowState(state: FlowState): OsqueryFlowListEntryBuilder {
+    this.flowListEntry.flow.state = state;
+    return this;
+  }
+
+  withQuery(query: string): OsqueryFlowListEntryBuilder {
+    this.query = query;
+    return this;
+  }
+
+  withStderr(stderr: string): OsqueryFlowListEntryBuilder {
+    this.stderr = stderr;
+    return this;
+  }
+
+  withResultsTable(columns: ReadonlyArray<string>, values: ReadonlyArray<ReadonlyArray<string>>): OsqueryFlowListEntryBuilder {
+    this.resultsTableBuilder
+      .withColumns(columns)
+      .withValues(values);
+    return this;
+  }
+
+  withProgressTable(
+    columns: ReadonlyArray<string>,
+    values: ReadonlyArray<ReadonlyArray<string>>
+  ): OsqueryFlowListEntryBuilder {
+    this.progressTableBuilder
+      .withColumns(columns)
+      .withValues(values);
+    return this;
+  }
+
+  withProgressRowsCount(count: number): OsqueryFlowListEntryBuilder {
+    this.progressRowsCount = count;
+    return this;
+  }
+
+  build(): FlowListEntry {
+    this.includeResultSet();
+    this.includeProgress();
+    this.setFlowArgsQuery(this.query);
+    return this.flowListEntry as FlowListEntry;
+  }
+
+  private setFlowArgsQuery(query: string): void {
+    if (this.flowListEntry.flow.args instanceof Object) {
+      this.flowListEntry.flow.args = {
+        ...this.flowListEntry.flow.args,
+        query,
+      };
+    } else {
+      this.flowListEntry.flow.args = { query };
+    }
+  }
+
+  private includeResultSet(): void {
+    const payload = {
+      stderr: this.stderr,
+      table: this.resultsTableBuilder.buildWithQueryIfDefined(this.query),
+    }
+    this.flowListEntry.resultSets = [
+      newFlowResultSet(payload) as DeepMutable<FlowResultSet>,
+    ];
+  }
+
+  private includeProgress(): void {
+    const progress = {
+      totalRowsCount: this.progressRowsCount,
+      partialTable: this.progressTableBuilder.buildWithQueryIfDefined(this.query),
+    };
+    this.flowListEntry.flow.progress = progress;
+  }
+}
+
+export function elementBySelector(selector: string, root: DebugElement): DebugElement {
```

Comment:
To me it's a bit awkward with or without the function, but I will remove it so that the code is consistent with the rest of the codebase. Also using `element.query(By.css('selector'))` is a general practice in testing, so I expect that the reader of the code knows what's going on and a separate function wouldn't add to the clarity too much.

For the record, I think there are two (subjective) benefits:
1. When we have many uses of it, `elementBySelector('foo bar', this.root)` stands out and the intent is clearer, whereas `this.root.query(By.css(selector))` initially looks like a random sequence of accessors (so it blends in with stuff like `.innerText` and others).
2. The `root` might be undefined in some cases, so we need optional chaining `?.`. This clutters the code a bit and there's some danger to miss it.